ntoskrnl/inbv/inbv.c

```diff
@@ -762,6 +769,147 @@ NtDisplayString(IN PUNICODE_STRING DisplayString)
     return STATUS_SUCCESS;
 }
 
+#ifdef INBV_ROTBAR_IMPLEMENTED
+static
+VOID
+NTAPI
+InbvRotationThread(
+    _In_ PVOID Context)
+{
+    ULONG X, Y, Index, Total = 18;
+    LARGE_INTEGER Delay;
+
+    InbvAcquireLock();
+    if (RotBarSelection == RB_SQUARE_CELLS)
+    {
+        Delay.QuadPart = -800000; // 0.08 sec
+        Index = 0;
+    }
+    else
+    {
+        Delay.QuadPart = -600000; // 0.06 sec
+        Index = 32;
+    }
+    X = ProgressBarLeft + 2;
+    Y = ProgressBarTop + 2;
+    InbvReleaseLock();
+
+    while (InbvDisplayState == INBV_DISPLAY_STATE_OWNED && PltRotBarStatus != 3)
```

Comment:
You need to hold the lock to check PltRotBarStatus. E.g. drop the release above and the acquire right below here, then add an acquire after the KeDelayExecutionThread and a release after the end of the loop. That way you'll be holding the lock when checking the loop condition and can't race with the stop function.

Reply:
Good idea!